sigar_linux.go

```diff
@@ -720,6 +723,35 @@ func (self *ProcTime) Get(pid int) error {
 	self.StartTime += system.btime
 	self.StartTime *= 1000
 
+	return err
+}
+
+// Diff the counters to calculate percent of CPU usage. The "other" ProcTime should be from an earlier reading.
+func (self *ProcTime) Diff(other *ProcTime) error {
+	// The diffs need to be protected against underflow
+	if other.User > self.User {
+		return errors.New("Failed diff user ProcTime: operation would result in underflow")
+	}
+	if other.Sys > self.Sys {
+		return errors.New("Failed diff sys ProcTime: operation would result in underflow")
+	}
+	if other.CollectionTime.After(self.CollectionTime) {
+		return errors.New("'Other' ProcTime should be older than current")
+	}
+
+	diffUser := uint64(self.User - other.User)
+	diffSys := uint64(self.Sys - other.Sys)
+	diffMillis := uint64(self.CollectionTime.Sub(other.CollectionTime) / time.Millisecond)
+
+	if diffMillis == 0 {
+		return errors.New("Failed to diff ProcTime: operation would result in divide by zero")
+	}
+
+	// Calculate percentages, multiplying by 100 first to avoid precision loss
+	self.PercentUserTime = (diffUser * 100) / diffMillis
+	self.PercentSysTime = (diffSys * 100) / diffMillis
+	self.PercentTotalTime = ((diffUser + diffSys) * 100) / diffMillis
```

Comment:
In the `Get()` method, the `User` and `Sys` times are converted to milliseconds of used CPU time. In that sense, the conversion has already taken place, and we just use the millisecond value. That's why we divide by `diffMillis`, to get the percentage of time used vs time passed. But absolutely the value of `system.ticks` corresponds to the `HZ` value found in `asm/param.h` or via `C.sysconf(C._SC_CLK_TCK)`, and `1/HZ` is a jiffy.

The tricky part is determining the value of `HZ` - its currently hardcoded to 100, yet we may someday run on platforms with a different value.